### PR TITLE
[WIP] Some fixes in NumericUpDown

### LINKED ISSUE #1885: Bug with Controls:NumericUpDown when setting Maximum programaticaly
I have a Flyout panel with some elements on it.
1. ComboBox
2. ComboBox
3. NumericUpDown -> Maximum is set programaticaly based on value selected in 2. Minimum is set to 0 in xaml.

If I go and select a value in 1. then go select a value in 2. and go and click on any of the +- signs in 3. , they are disabled. I fix this by entering a number manualy, then going and clicking on another control (1. or 2.) then coming back to 3. Arrows are now finaly enabled. This doesn't happen if I set the Maximum only from XAML. I tried setting both in xaml and again in codebehind but issue still occurs

EDIT: I also tryied binding it from XAML to a int in codebehind, issue still occurs.

### PR DESCRIPTION
**Pull request by xxMUROxx**
- [x] NumericUpDown Hexadecimal Issue #1375 
- [x] NumericUpDown: UseFloatingWatermark #2202
- [x] Remove duplicate code in *SetTextLength* and *XXX-GotFocus*
- [ ] Fixes #1885 *(should have been fixed), need feedback from issue-creator*

### CODE REVIEW

**Comment by xxMUROxx** [comment]
Please wait before merging.

**Comment by xxMUROxx** [comment]
@punker76: I think this can be merged.
I have refactored a bit the TextBoxHelper class to avoid duplicate code. The content in the *Files changed* tab looks a bit confusing but it should not break anything.
Please double-check if there are any weird behaviors.

File: MahApps.Metro/Controls/Helper/TextBoxHelper.cs [hunk 1]
```diff
@@ -41,6 +41,14 @@ public class TextBoxHelper
...
         private static readonly DependencyProperty IsSpellCheckContextMenuEnabledProperty = DependencyProperty.RegisterAttached("IsSpellCheckContextMenuEnabled", typeof(bool), typeof(TextBoxHelper), new FrameworkPropertyMetadata(false, UseSpellCheckContextMenuChanged));
 
+        private static readonly Func<NumericUpDown, int> NumericUpDownTextLength = control => control.Value.HasValue ? 1 : 0;
+        private static readonly Func<PasswordBox, int> PasswordBoxTextLength = control => control.Password.Length;
+        private static readonly Func<TextBox, int> TextBoxTextLength = control => control.Text.Length;
+
+        private static readonly Func<NumericUpDown, Action> NumericUpDownFocus = control => control.SelectAll;
+        private static readonly Func<PasswordBox, Action> PasswordBoxFocus = control => control.SelectAll;
+        private static readonly Func<TextBox, Action> TextBoxFocus = control => control.SelectAll;
+
```
Comment (punker76) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @xxMUROxx do we really need this `static` stuff? cause every func is only used once...

Comment (xxMUROxx) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @punker76 we could also inline this fields. But what do you think of both generic methods? Should simplify further development if we add *UseFloatingWatermark* for other controls.

Comment (punker76) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: >But what do you think of both generic methods? Should simplify further development if we add UseFloatingWatermark for other controls.

:+1:

Comment (xxMUROxx) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @punker76 done

File: MahApps.Metro/Themes/NumericUpDown.xaml [hunk 2]
```diff
@@ -68,8 +68,31 @@
...
+                                           Foreground="{TemplateBinding Foreground}"
+                                           Opacity="0.6"
+                                           HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}"
+                                           VerticalAlignment="{TemplateBinding VerticalContentAlignment}">
+                                    <TextBlock.RenderTransform>
+                                        <TranslateTransform />
+                                    </TextBlock.RenderTransform>
+                                </TextBlock>
+                            </Grid>
                             <TextBox x:Name="PART_TextBox"
```
Comment (punker76) [inline MahApps.Metro/Themes/NumericUpDown.xaml]: @xxMUROxx can you change the floating message part like in https://github.com/MahApps/MahApps.Metro/commit/1919458bce35ba5c5815e2c9982721836263cdb2 ?
thx

Comment (xxMUROxx) [inline MahApps.Metro/Themes/NumericUpDown.xaml]: @punker76 do you mean removing *PART_FloatingMessage*

Comment (punker76) [inline MahApps.Metro/Themes/NumericUpDown.xaml]: @xxMUROxx no, i mean changeing Grid -> ContentControl and the tirgger part

https://github.com/MahApps/MahApps.Metro/commit/1919458bce35ba5c5815e2c9982721836263cdb2#diff-e09aa3d83a4c174f3c72c7698ff3c40eR124

https://github.com/MahApps/MahApps.Metro/commit/1919458bce35ba5c5815e2c9982721836263cdb2#diff-e09aa3d83a4c174f3c72c7698ff3c40eR193

**Comment by xxMUROxx** [comment]
@punker76  :+1:

**Pull request merged**